idaes/gas_solid_contactors/properties/oxygen_iron_OC_oxidation/gas_phase_thermo.py

```diff
@@ -0,0 +1,798 @@
+
+        # ---------------------------------------------------------------------
+        # Initialise values
+        for k in blk.keys():
+
+            if hasattr(blk[k], "mw_eqn"):
+                calculate_variable_from_constraint(
+                            blk[k].mw,
+                            blk[k].mw_eqn)
+
+            if hasattr(blk[k], "ideal_gas"):
+                calculate_variable_from_constraint(
+                            blk[k].dens_mol,
+                            blk[k].ideal_gas)
+
+            if hasattr(blk[k], "dens_mass_basis"):
+                calculate_variable_from_constraint(
+                            blk[k].dens_mass,
+                            blk[k].dens_mass_basis)
+
+            if hasattr(blk[k], "mixture_heat_capacity_eqn"):
+                calculate_variable_from_constraint(
+                            blk[k].cp_mol,
+                            blk[k].mixture_heat_capacity_eqn)
+
+            if hasattr(blk[k], "cp_mass_basis"):
+                calculate_variable_from_constraint(
+                            blk[k].cp_mass,
+                            blk[k].cp_mass_basis)
+
+            if hasattr(blk[k], "visc_d_constraint"):
+                calculate_variable_from_constraint(
+                            blk[k].visc_d,
+                            blk[k].visc_d_constraint)
+
+            if hasattr(blk[k], "therm_cond_constraint"):
+                calculate_variable_from_constraint(
+                            blk[k].therm_cond,
+                            blk[k].therm_cond_constraint)
+
+            if hasattr(blk[k], "mixture_enthalpy_eqn"):
+                calculate_variable_from_constraint(
+                            blk[k].enth_mol,
+                            blk[k].mixture_enthalpy_eqn)
+
+            for j in blk[k]._params.component_list:
+
+                if hasattr(blk[k], "comp_conc_eqn"):
+                    calculate_variable_from_constraint(
+                                blk[k].dens_mol_comp[j],
+                                blk[k].comp_conc_eqn[j])
+
+                if hasattr(blk[k], "diffusion_comp_constraint"):
+                    calculate_variable_from_constraint(
+                                blk[k].diffusion_comp[j],
+                                blk[k].diffusion_comp_constraint[j])
+
+                if hasattr(blk[k], "cp_shomate_eqn"):
+                    calculate_variable_from_constraint(
+                        blk[k].cp_mol_comp[j],
+                        blk[k].cp_shomate_eqn[j])
+
+                if hasattr(blk[k], "enthalpy_shomate_eqn"):
+                    calculate_variable_from_constraint(
+                            blk[k].enth_mol_comp[j],
+                            blk[k].enthalpy_shomate_eqn[j])
+
+        # Solve property block if non-empty
+        free_vars = 0
+        for k in blk.keys():
+            free_vars += number_unfixed_variables_in_activated_equalities(
+                blk[k])
```

Comment:
What if the number of free vars in each key are different? Depending on what property methods are called from the unit model, this might be a possibility.

Reply:
The issue here is that in some cases a state block can end up with no unfixed variables (and thus no constraints), which causes an error in IPOPT.

One alternative solution would be to only search until an unfixed variable is found and then proceed - that would probably be more efficient as you would probably either find an unfixed variable fairly quickly, or not have many variables to iterate over in the case of a model with no unfixed vars.

That said, there might be better solutions than that. Do you have any suggestions @Robbybp?

Reply:
Searching only until a single variable is found sounds like a good idea. Really, I would like to omit this full-space nonlinear solve if it is unnecessary, but determining this at run-time would require a block triangularization. I don't think this is prohibitive, and I don't think it should be any more expensive than writing the nl file, but I haven't settled on what I think is the best way to implement this.